### PR TITLE
build-configs-android: enable clang-14 android builds

### PR DESCRIPTION
**Pull request by musamaanjum**
Enable clang-14 builds on 5.10 and 5.15 android branches. Also enable
the GKI configs.

Signed-off-by: Muhammad Usama Anjum <usama.anjum@collabora.com>

### CODE REVIEW

File: config/core/build-configs-android.yaml
Comment (gctucker) [inline config/core/build-configs-android.yaml:62]: It's a bit better to inherit from the definition in `gcc-10` and override fields that change, for example:
```suggestion
        x86_64:
          <<: * x86_64_arch
          extra_configs:
            - 'allmodconfig'
            - 'allnoconfig'
            - 'gki_config'
```
The same can be done for `arm64` even though there's nothing to inherit.  It makes the config easier to maintain as future changes in the `gcc-10` section that are architecture-specific would be automatically propagated to the other sections such as `clang-14`.

**Review by gctucker — COMMENTED** [review]
LGTM, let's see if we can get some results on staging.

If some builds aren't working at all, we may have to drop them from this PR and deal with them separately (if additional compiler options are needed etc.).  But that will depend on the results.

**Comment by gctucker** [comment]
I wonder if all the kernel branches should be built with `clang-14` or even with Clang at all, the list goes back to 4.4.  It might be worth asking on IRC, and if we need to only enable recent kernels with `clang-14` then we could create an `android_variants_clang` definition to use with those branches only.

**Comment by gctucker** [comment]
Same thing for `gki_config` - in which kernel revision were they introduced?

**Comment by musamaanjum** [comment]
I'd asked on IRC yesterday. I'm waiting for reply:
> In kernelCI, we have android kernels which date back to `4.4`. GKI configs for android are present in android-4.19-stable, android11-5.4 and so on. We should enable the android builds with clang-14 only on recent kernels. Which kernels seem good candidates to be enabled on clang-14? kernels 5.4 onward seem good candidates. Android's currently enabled branches can be found here: config/core/build-configs-android.yaml

**Comment by gctucker** [comment]
OK it's probably also possible to look for the presence of the `gki_config` files in each branch.

**Comment by gctucker** [comment]
Staging builds are here:
https://staging.kernelci.org/build/android/branch/android13-5.15/kernel/ASB-2021-12-05_13-5.15-461-gf4bb4932c9ec/

In particular:
FAIL https://staging.kernelci.org/build/id/61c9be82a2f2f74e6363557b/ `x86_64_defconfig`
FAIL https://staging.kernelci.org/build/id/61c9b2fca6ca40fe9663557c/ `multi_v7_defconfig`

The `defconfig` builds for `riscv` and `arm64` passed.  We should check whether these build errors are expected (i.e. known problems in the kernel tree) or if they are due to KernelCI not building correctly.

**Comment by musamaanjum** [comment]
All the failing builds [here](https://staging.kernelci.org/build/android/branch/android13-5.15/kernel/ASB-2021-12-05_13-5.15-461-gf4bb4932c9ec/) will be solved by the already identified two patches except the following builds:
[ixp4xx_defconfig ‐ arm ‐ clang-14](https://storage.staging.kernelci.org/android/android13-5.15/asb-2021-12-05_13-5.15-461-gf4bb4932c9ec/arm/ixp4xx_defconfig/clang-14/logs/build-errors.log)
[rpc_defconfig ‐ arm ‐ clang-14](https://storage.staging.kernelci.org/android/android13-5.15/asb-2021-12-05_13-5.15-461-gf4bb4932c9ec/arm/rpc_defconfig/clang-14/logs/build-errors.log)
[pxa_defconfig ‐ arm ‐ clang-14](https://storage.staging.kernelci.org/android/android13-5.15/asb-2021-12-05_13-5.15-461-gf4bb4932c9ec/arm/pxa_defconfig/clang-14/logs/build-errors.log)
These are failing with different build errors.

**Comment by gctucker** [comment]
> All the failing builds [here](https://staging.kernelci.org/build/android/branch/android13-5.15/kernel/ASB-2021-12-05_13-5.15-461-gf4bb4932c9ec/) will be solved by the already identified two patches except the following builds: [ixp4xx_defconfig ‐ arm ‐ clang-14](https://storage.staging.kernelci.org/android/android13-5.15/asb-2021-12-05_13-5.15-461-gf4bb4932c9ec/arm/ixp4xx_defconfig/clang-14/logs/build-errors.log) [rpc_defconfig ‐ arm ‐ clang-14](https://storage.staging.kernelci.org/android/android13-5.15/asb-2021-12-05_13-5.15-461-gf4bb4932c9ec/arm/rpc_defconfig/clang-14/logs/build-errors.log) [pxa_defconfig ‐ arm ‐ clang-14](https://storage.staging.kernelci.org/android/android13-5.15/asb-2021-12-05_13-5.15-461-gf4bb4932c9ec/arm/pxa_defconfig/clang-14/logs/build-errors.log) These are failing with different build errors.

OK, thanks for checking.  I don't think this will be a blocking issue but we should probably review the list of defconfigs being built with Android branches anyway.  Maybe we can stop building all the `arm` ones specific to a particular platform, it's something to confirm with Google.  It would save resources and make the results land quicker.

**Comment by gctucker** [comment]
Confirmed on staging that all clang-14 builds are passing on linux-next with the extra fix applied:
https://staging.kernelci.org/build/kernelci/branch/staging-next/kernel/staging-next-20211228.1/

**Comment by musamaanjum** [comment]
After discussing with @gctucker, I've updated the PR to enable only 5.10 and 5.15 andorid branches with clang 14. GKI config is also present on these branches. We can always include more android branches after consulting Google.

**Review by gctucker — APPROVED** [review]
LGTM, starting a build job on staging to check it's producing valid results.

**Comment by musamaanjum** [comment]
Checked the results: https://staging.kernelci.org/build/android/branch/android13-5.15/kernel/ASB-2022-01-05_13-5.15-55-g9e280ea43eaa/

Other than builds already mentioned [failing builds above](https://github.com/kernelci/kernelci-core/pull/974#issuecomment-1002433487), everything should be good after we add the [previously identified 2 patches](https://github.com/kernelci/kernelci-core/issues/489#issuecomment-1001963194) to android stagging as well.

**Review by gctucker — APPROVED** [review]
Tested OK on staging.

**Pull request merged**
